system-addon/lib/TelemetryFeed.jsm

```diff
@@ -0,0 +1,162 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+/* globals XPCOMUtils, gUUIDGenerator, ClientID */
+
+"use strict";
+
+const {utils: Cu} = Components;
+const {actionTypes: at, actionUtils: au} = Cu.import("resource://activity-stream/common/Actions.jsm", {});
+
+Cu.import("resource://gre/modules/ClientID.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyServiceGetter(this, "gUUIDGenerator",
+  "@mozilla.org/uuid-generator;1",
+  "nsIUUIDGenerator");
+XPCOMUtils.defineLazyModuleGetter(this, "TelemetrySender",
+  "resource://activity-stream/lib/TelemetrySender.jsm");
+
+this.TelemetryFeed = class TelemetryFeed {
+  constructor(options) {
+    this.sessions = new Map();
+    this.telemetryClientId = null;
+    this.telemetrySender = null;
+  }
+
+  async init() {
+    // TelemetrySender adds pref observers, so we initialize it after INIT
+    this.telemetrySender = new TelemetrySender();
+
+    const id = await ClientID.getClientID();
+    this.telemetryClientId = id;
+  }
+
+  /**
+   * addSession - Start tracking a new session
+   *
+   * @param  {string} id the portID of the open session
+   */
+  addSession(id) {
+    this.sessions.set(id, {
+      start_time: Components.utils.now(),
+      session_id: String(gUUIDGenerator.generateUUID()),
+      page: "NEW_TAB" // TODO: Handle about:home
+    });
+  }
+
+  /**
+   * endSession - Stop tracking a session
+   *
+   * @param  {string} portID the portID of the session that just closed
+   */
+  endSession(portID) {
+    const session = this.sessions.get(portID);
+
+    if (!session) {
+      // It's possible the tab was never visible – in which case, there was no user session.
+      return;
+    }
+
+    session.session_duration = Math.round(Components.utils.now() - session.start_time);
+    this.sendEvent(this.createSessionEndEvent(session));
+    this.sessions.delete(portID);
+  }
+
+  /**
+   * createPing - Create a ping with common properties
+   *
+   * @param  {string} id The portID of the session
+   * @return {obj}    A telemetry ping
+   */
+  createPing(portID) {
+    const appInfo = this.store.getState().App;
+    const ping = {
+      client_id: this.telemetryClientId,
+      addon_version: appInfo.version,
+      locale: appInfo.locale
+    };
+
+    // If the ping is part of a user session, add session-related info
+    if (portID) {
+      const session = this.sessions.get(portID);
+      Object.assign(ping, {
+        session_id: session.session_id,
+        page: session.page
+      });
+    }
+    return ping;
+  }
+
+  createUserEvent(action) {
+    return Object.assign(
+      this.createPing(au.getPortIdOfSender(action)),
+      action.data,
+      {action: "activity_stream_event"}
+    );
+  }
+
+  createUndesiredEvent(action) {
+    return Object.assign(
+      this.createPing(au.getPortIdOfSender(action)),
+      {value: 0}, // Default value
+      action.data,
+      {action: "activity_stream_masga_event"}
```

Comment:
I agree, I think that would be something we have to change on the data pipeline side though (or at least talk to marina about)